File: api/app/database/database-singleton.js
const { CosmosClient } = require('@azure/cosmos')
const config = require('../config/config')

class CosmosDBService {
  static cosmosClient = null
  static database = null
  static containers = {}

  static async getInstance() {
    if (!this.cosmosClient) {
      await this.initializeCosmosClient()
    }
    return this
  }

  static async initializeCosmosClient() {
    const endpoint = config.db.cosmos_endpoint
    const key = config.db.cosmos_key
    const databaseId = config.db.cosmos_db_id

    this.cosmosClient = new CosmosClient({ endpoint, key })

    const { database } = await this.cosmosClient.databases.createIfNotExists({ id: databaseId })
    this.database = database

    await this.initializeContainers([
      'Incidents',
      'Locations',
      'Resources',
      'ResponsePlans',
      'Responses',
    ])
  }

  static async initializeContainers(containerIds) {
    for (const id of containerIds) {
      const { container } = await this.database.containers.createIfNotExists({ id })
      this.containers[id] = container
    }
  }

  static async queryItems(query, containerId) {
    if (!this.containers[containerId]) {
      throw new Error(`Container ${containerId} does not exist or has not been initialized.`)
    }

    const container = this.containers[containerId]
    const { resources: items } = await container.items.query(query).fetchAll()
    return items
  }

  static async addItem(item, containerId) {
    if (!this.containers[containerId]) {
      throw new Error(`Container ${containerId} does not exist or has not been initialized.`)
    }

    const container = this.containers[containerId]
    const { resource: addedItem } = await container.items.create(item)
    return addedItem
  }

  static async updateItem(itemId, newItem, containerId) {
    if (!this.containers[containerId]) {
      throw new Error(`Container ${containerId} does not exist or has not been initialized.`)
    }
    const container = this.containers[containerId]
    const { resource: updatedItem } = await container
      .item(itemId, newItem.partitionKey)
      .replace(newItem)
    return updatedItem
  }

  static async deleteItem(itemId, partitionKey, containerId) {
    if (!this.containers[containerId]) {
      throw new Error(`Container ${containerId} does not exist or has not been initialized.`)
    }
    const container = this.containers[containerId]
    await container.item(itemId, partitionKey).delete()
  }
}

module.exports = CosmosDBService

File: hapi/app/services/database_singleton.py
from azure.cosmos import CosmosClient, PartitionKey, exceptions
from flask import current_app

class CosmosDBService:
    cosmos_client = None
    database = None
    containers = {}

    @classmethod
    def get_instance(cls):
        if cls.cosmos_client is None:
            cls.initialize_cosmos_client()
        return cls

    @classmethod
    def initialize_cosmos_client(cls):
        connection_string = current_app.config['COSMOS_CONNECTION_STRING']
        database_id = current_app.config['COSMOS_DB_ID']

        cls.cosmos_client = CosmosClient.from_connection_string(connection_string)
        cls.database = cls.cosmos_client.create_database_if_not_exists(id=database_id)

        cls.initialize_containers([
            'Incidents',
            'Locations',
            'Resources',
            'ResponsePlans',
            'Responses',
        ])

    @classmethod
    def initialize_containers(cls, container_ids):
        for container_id in container_ids:
            container = cls.database.create_container_if_not_exists(
                id=container_id,
                partition_key=PartitionKey(path="/id")  # Adjust if you use another partition key
            )
            cls.containers[container_id] = container

    @classmethod
    def query_items(cls, query, container_id):
        if container_id not in cls.containers:
            raise Exception(f"Container {container_id} does not exist or has not been initialized.")

        container = cls.containers[container_id]
        items = list(container.query_items(query=query, enable_cross_partition_query=True))
        return items

    @classmethod
    def add_item(cls, item, container_id):
        if container_id not in cls.containers:
            raise Exception(f"Container {container_id} does not exist or has not been initialized.")

        container = cls.containers[container_id]
        added_item = container.create_item(body=item)
        return added_item

    @classmethod
    def update_item(cls, item_id, new_item, container_id):
        if container_id not in cls.containers:
            raise Exception(f"Container {container_id} does not exist or has not been initialized.")

        container = cls.containers[container_id]
        updated_item = container.replace_item(item=item_id, body=new_item)
        return updated_item

    @classmethod
    def delete_item(cls, item_id, partition_key, container_id):
        if container_id not in cls.containers:
            raise Exception(f"Container {container_id} does not exist or has not been initialized.")

        container = cls.containers[container_id]
        container.delete_item(item=item_id, partition_key=partition_key)
